packages/bsky/src/views/threadsV2.ts

```diff
@@ -0,0 +1,294 @@
+import { AppBskyUnspeccedGetPostThreadV2, asPredicate } from '@atproto/api'
+import { HydrateCtx } from '../hydration/hydrator'
+import { validateRecord as validatePostRecord } from '../lexicon/types/app/bsky/feed/post'
+import {
+  QueryParams as GetPostThreadV2QueryParams,
+  ThreadItem,
+  ThreadItemBlocked,
+  ThreadItemNoUnauthenticated,
+  ThreadItemNotFound,
+  ThreadItemPost,
+} from '../lexicon/types/app/bsky/unspecced/getPostThreadV2'
+import { $Typed } from '../lexicon/util'
+
+type ThreadItemValue<T extends ThreadItem['value']> = Omit<
+  ThreadItem,
+  'value'
+> & {
+  value: T
+}
+
+export type ThreadItemValueBlocked = ThreadItemValue<$Typed<ThreadItemBlocked>>
+
+export type ThreadItemValueNoUnauthenticated = ThreadItemValue<
+  $Typed<ThreadItemNoUnauthenticated>
+>
+
+export type ThreadItemValueNotFound = ThreadItemValue<
+  $Typed<ThreadItemNotFound>
+>
+
+export type ThreadItemValuePost = ThreadItemValue<$Typed<ThreadItemPost>>
+
+type ThreadBlockedNode = {
+  item: ThreadItemValueBlocked
+}
+type ThreadNoUnauthenticatedNode = {
+  parent: ThreadTree | undefined
+  item: ThreadItemValueNoUnauthenticated
+}
+
+type ThreadNotFoundNode = {
+  item: ThreadItemValueNotFound
+}
+
+type ThreadPostNode = {
+  item: ThreadItemValuePost
+  hasOPLike: boolean
+  parent: ThreadTree | undefined
+  replies: ThreadTree[] | undefined
+}
+
+const isThreadNoUnauthenticatedNode = (
+  node: ThreadTree,
+): node is ThreadNoUnauthenticatedNode =>
+  AppBskyUnspeccedGetPostThreadV2.isThreadItemNoUnauthenticated(node.item.value)
+
+const isThreadPostNode = (node: ThreadTree): node is ThreadPostNode =>
+  AppBskyUnspeccedGetPostThreadV2.isThreadItemPost(node.item.value)
+
+export type ThreadTree =
+  | ThreadBlockedNode
+  | ThreadNoUnauthenticatedNode
+  | ThreadNotFoundNode
+  | ThreadPostNode
+
+export function sortTrimFlattenThreadTree(
+  anchorTree: ThreadTree,
+  options: SortTrimFlattenOptions,
+) {
+  const sortedAnchorTree = sortTrimThreadTree(anchorTree, options)
+  return flattenThree(sortedAnchorTree)
+}
+
+type SortTrimFlattenOptions = {
+  branchingFactor: GetPostThreadV2QueryParams['branchingFactor']
+  fetchedAt: number
+  opDid: string
+  prioritizeFollowedUsers: boolean
+  sorting: GetPostThreadV2QueryParams['sorting']
+  viewer: HydrateCtx['viewer']
+}
+
+const isPostRecord = asPredicate(validatePostRecord)
+
+function sortTrimThreadTree(
+  node: ThreadTree,
+  opts: SortTrimFlattenOptions,
+): ThreadTree {
+  if (!isThreadPostNode(node)) {
+    return node
+  }
+
+  const {
+    branchingFactor,
+    fetchedAt,
+    opDid,
+    prioritizeFollowedUsers,
+    sorting,
+    viewer,
+  } = opts
+
+  if (node.replies) {
+    node.replies.sort((aTree: ThreadTree, bTree: ThreadTree) => {
```

Comment:
I'm not opposed to this function `sortTrimThreadTree()` mutating its input `node`, but I think it would be nice to call out at the top of the function as a headsup.